### PR TITLE
feat: add graceful shutdown option for apply&destroy

### PR DESCRIPTION
**Pull request by wjf3121**
The default behavior of `CommandContext` is to [immediately kill](https://cs.opensource.google/go/go/+/master:src/os/exec/exec.go;l=453;drc=1e12eab8705d1d8d7472be9147a39caa1c8380db) the command process when receiving cancellation signal from the context, which may lead to corrupted terraform state. 

This PR adds the option to allow sending interruption signal to the Apply/Destroy command and wait for a graceful period before forcefully killing the process. Note that the option doesn't work on Windows due to [the lack of support](https://cs.opensource.google/go/go/+/master:src/os/exec.go;l=138-139;drc=73847aa41b9299a39e596651ccdd9c0dcb511d5d) of sending interrupt signal.

Test: all ut and e2e passed, verified the behavior with RWC CLI

### CODE REVIEW

File: tfexec/cmd.go [hunk 1]
```diff
@@ -275,3 +303,97 @@ func writeOutput(ctx context.Context, r io.ReadCloser, w io.Writer) error {
 		}
 	}
 }
+
+func (tf *Terraform) runTerraformCmd(ctx context.Context, cmd *exec.Cmd) error {
```
Comment (wjf3121) [inline tfexec/cmd.go:305]: Merged the runTerraformCmd implementation in `cmd_default.go` an `cmd_linux.go` into one. The only difference is the different value for `cmd.SysProcAttr` at line 310.

File: tfexec/cmd_default.go [hunk 2]
```diff
@@ -1,95 +1,11 @@
...
-import (
-	"context"
-	"fmt"
-	"os/exec"
-	"strings"
-	"sync"
-)
+import "syscall"
 
-func (tf *Terraform) runTerraformCmd(ctx context.Context, cmd *exec.Cmd) error {
```
Comment (wjf3121) [inline tfexec/cmd_default.go:17]: This is moved to `cmd.go`.

File: tfexec/cmd_linux.go [hunk 3]
```diff
@@ -1,100 +1,16 @@
...
-import (
-	"context"
-	"fmt"
-	"os/exec"
-	"strings"
-	"sync"
-	"syscall"
-)
-
-func (tf *Terraform) runTerraformCmd(ctx context.Context, cmd *exec.Cmd) error {
```
Comment (wjf3121) [inline tfexec/cmd_linux.go:15]: This is moved to `cmd.go`.

File: go.mod [hunk 4]
```diff
@@ -1,6 +1,6 @@
 module github.com/hashicorp/terraform-exec
 
-go 1.18
+go 1.20
```
Comment (wjf3121) [inline go.mod:3]: This is the least version supporting `Cancel` and ` WaitDelay`: https://tip.golang.org/doc/go1.20#minor_library_changes

File: tfexec/cmd_test.go [hunk 5]
```diff
@@ -100,3 +104,29 @@ func assertCmd(t *testing.T, expectedArgs []string, expectedEnv map[string]strin
 		}
 	}
 }
+
+func TestRunTerraformCmd(t *testing.T) {
```
Comment (wjf3121) [inline tfexec/cmd_test.go:108]: This is also a merged from `cmd_default_test.go` and `cmd_linux_test.go`.

File: tfexec/cmd.go [hunk 6]
```diff
@@ -275,3 +303,97 @@ func writeOutput(ctx context.Context, r io.ReadCloser, w io.Writer) error {
...
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		errStdoutCtx := ctx
+		// When cancellation delay is enabled, avoid piping cancellation to prevent
+		// cancellation from being interrupted due to broken pipe.
+		if cmd.WaitDelay > 0 {
+			errStdoutCtx = context.WithoutCancel(ctx)
+		}
+		errStdout = writeOutput(errStdoutCtx, stdoutPipe, stdoutWriter)
```
Comment (Gogomoe) [inline tfexec/cmd.go:358]: question: Where do we wait for the subprocess exit? here (wait for pipe closed)?

File: tfexec/cmd.go [hunk 7]
```diff
@@ -193,6 +195,32 @@ func (tf *Terraform) buildTerraformCmd(ctx context.Context, mergeEnv map[string]
...
 
+func (tf *Terraform) buildTerraformCmdWithGracefulShutdown(ctx context.Context, gracefulShutdownPeriod time.Duration, mergeEnv map[string]string, args ...string) *exec.Cmd {
+	cmd := exec.CommandContext(ctx, tf.execPath, args...)
+
+	cmd.Env = tf.buildEnv(mergeEnv)
+	cmd.Dir = tf.workingDir
+
+	if gracefulShutdownPeriod > 0 {
+		tf.logger.Printf("[INFO] configuring gracefully cancellation for command: %s", cmd.String())
+		cmd.WaitDelay = gracefulShutdownPeriod
```
Comment (Gogomoe) [inline tfexec/cmd.go]: question: I don't know much about the `cmd.WaitDelay`. I guess when we send `os.Interrupt` to the subprocess, the subprocess must exit in `cmd.WaitDelay`. Does my understanding right?

Comment (Gogomoe) [inline tfexec/cmd.go]: question: If the terraform process doesn't exit in time after sending `os.Interrupt`. Is this case exist? where do we handle it?

File: tfexec/cmd.go (hunk 6 above)
Comment (wjf3121) [reply tfexec/cmd.go:358]: We have `wg.Wait()` at line 377 to wait for the buffer writing to finish. Once the graceful cancellation pass the timeout period, The command be sent kill signal, so this is guaranteed to be finished.

Does this answer your question?

File: tfexec/cmd.go (hunk 7 above)
Comment (wjf3121) [reply tfexec/cmd.go]: Pretty much, after `cmd.WaitDelay`, the standard library will send kill signal for force quitting the subprocess. This is the corresponding implementation: https://cs.opensource.google/go/go/+/master:src/os/exec/exec.go;l=793-813;drc=a548dee75cc2fd578acc303f58887a50e64e29bc

Comment (wjf3121) [reply tfexec/cmd.go]: Then it will be force killed and we will still have a chance to experience data corruption.

Comment (wjf3121) [reply tfexec/cmd.go]: We can also add an option to allow infinite waiting by setting `cmd.WaitDelay` to zero, then the program will wait indefinitely until the subprocess finishes and close its pipe.

Let me know what you think.

Comment (wjf3121) [reply tfexec/cmd.go]: I updated the API to allow zero `cmd.WaitDelay`. PTAL

**Review by Gogomoe — APPROVED** [review]
LGTM!

**Pull request merged**